Add shortname ambiguity warning in shortcut expander

ardaguclu opened the pull request:
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
Currently, errors in client-go are propagated back to the callers via
function returns. However, there is no elegant way for just warning users.
For example, when user wants to get a resource with it's short name format
and if there are multiple resources belonging to this short name, we need to
warn user about this ambiguity which one is picked and which ones are discarded.

Not only to overcome this particular case mentioned above, but also propose a
way for the possible warnings in the future, this PR adds a warningHandler
callback function in shortcutExpander as well as other necessary bindings from kubectl
-> cli-runtime -> client go

#### Which issue(s) this PR fixes:
Fixes https://github.com/kubernetes/kubernetes/issues/108573

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

ardaguclu committed:
Add warning handler callback function in shortcut expander

Currently, errors in client-go are propagated back to the callers via
function returns. However, there is no elegant way for just warning users.
For example, when user wants to get a resource with it's short name format
and if there are multiple resources belonging to this short name, we need to
warn user about this ambugity which one is picked and which ones are discarded.

Not only to overcome this particular case mentioned above, but also propose a
way for the possible warnings in the future, this commit adds a warningHandler
callback function in shortcutExpander.

ardaguclu commented:
/cc @deads2k @liggitt 
would you mind taking a look at this one?. Thanks.

ardaguclu commented:
/triage accepted
/priority backlog

liggitt commented on staging/src/k8s.io/client-go/restmapper/shortcut.go:
```diff
@@ -145,16 +145,25 @@ func (e shortcutExpander) expandResourceShortcut(resource schema.GroupVersionRes
-				resource.Group = item.LongForm.Group
-				return resource
+				if found {
+					klog.V(1).Infof("Skipping ambiguous resource %q for %s. Continuing with %s. Please specify the fully qualified resource name.", item.LongForm, resource, rsc)
```
a few thoughts:
1. This is a really long message that could be repeated multiple times... an example output when this happens would make this easier to see. It seems like some collapsing/combining should be done.
2. klog-formatted output doesn't seem like a great user experience (prefixing with shortcut.go / timestamp / etc)
3. adding logging into library methods rather than propagating the information back to the caller seems to go against the spirit of https://github.com/kubernetes/kubernetes/issues/94428

liggitt replied:
as an alternative, NewShortcutExpander or an alternate constructor could take some input/indication of what to do with ambiguous resources

ardaguclu replied:
@liggitt would you mind taking a look at the latest version of this PR?. Thanks.

ardaguclu committed:
Add warningPrinter functionality in ConfigFlags

ConfigFlags has neither warning user in a standardized
format functionality nor passing warning callback functions to other upper level
libraries such as client-go.

This commit adds an ability that user can set warningPrinters
according to their IOStreams and this warningPrinters will be used
to raise possible warnings happening not only in cli-runtime but
also in client-go.

ardaguclu committed:
Pass warning callback function in ConfigFlags to shortcutExpander

This commit passes warning callback function to print possible
warnings happened in shortcut expander to warn user in a
standardized format.

ardaguclu committed:
Add integration test for CRDs having ambiguous short names

This commit adds integration test to assure that warning message
related to this ambiguity is printed when resources are being retrieved via their short name
representations in cases where multiple resources have same
short names.

This integration test also ensures that the logic behind which resource
will be selected hasn't been changed which may cause disperancies in
clusters.

ardaguclu commented:
/test pull-kubernetes-e2e-gce

apelisse commented on staging/src/k8s.io/cli-runtime/pkg/printers/terminal.go:
```diff
@@ -37,3 +41,35 @@ func WriteEscaped(writer io.Writer, output string) error {
+		return false
+	}
+
+	// https://no-color.org/
+	if _, nocolor := os.LookupEnv("NO_COLOR"); nocolor {
+		return false
+	}
```
Thanks for that!

liggitt commented on staging/src/k8s.io/client-go/restmapper/shortcut.go:
```diff
@@ -145,16 +148,27 @@ func (e shortcutExpander) expandResourceShortcut(resource schema.GroupVersionRes
-				return resource
+				if found {
+					if e.warningHandler != nil {
+						e.warningHandler(fmt.Sprintf("ambiguous resource %q for %s is skipped", item.LongForm, resource))
```
does this print warnings if multiple versions of the same group/resource all register the same shortname?

ardaguclu replied:
I tried this with example CRD;

```yaml
apiVersion: apiextensions.k8s.io/v1
kind: CustomResourceDefinition
metadata:
  # name must match the spec fields below, and be in the form: <plural>.<group>
  name: crontabs.example.com
spec:
  # group name to use for REST API: /apis/<group>/<version>
  group: example.com
  # list of versions supported by this CustomResourceDefinition
  versions:
  - name: v1beta1
    # Each version can be enabled/disabled by Served flag.
    served: true
    # One and only one version must be marked as the storage version.
    storage: true
    # A schema is required
    schema:
      openAPIV3Schema:
        type: object
        properties:
          host:
            type: string
          port:
            type: string
  - name: v1
    served: true
    storage: false
    schema:
      openAPIV3Schema:
        type: object
        properties:
          host:
            type: string
          port:
            type: string
  # The conversion section is introduced in Kubernetes 1.13+ with a default value of
  # None conversion (strategy sub-field set to None).
  conversion:
    # None conversion assumes the same schema for all versions and only sets the apiVersion
    # field of custom resources to the proper value
    strategy: None
  # either Namespaced or Cluster
  scope: Namespaced
  names:
    # plural name to be used in the URL: /apis/<group>/<version>/<plural>
    plural: crontabs
    # singular name to be used as an alias on the CLI and for display
    singular: crontab
    # kind is normally the CamelCased singular type. Your resource manifests use this.
    kind: CronTab
    # shortNames allow shorter string to match your resource on the CLI
    shortNames:
    - ct
```

and when I called `kubectl get ct`, it showed warning. I think, it should not warn this cases?

liggitt replied:
> I think, it should not warn this cases?

I agree

liggitt commented on staging/src/k8s.io/kubectl/pkg/cmd/cmd.go:
```diff
@@ -100,11 +100,13 @@ var defaultConfigFlags = genericclioptions.NewConfigFlags(true).WithDeprecatedPa
 // NewDefaultKubectlCommand creates the `kubectl` command with default arguments
 func NewDefaultKubectlCommand() *cobra.Command {
+	ioStreams := genericiooptions.IOStreams{In: os.Stdin, Out: os.Stdout, ErrOut: os.Stderr}
+	defaultConfigFlags = defaultConfigFlags.WithWarningPrinter(ioStreams)
```
this is mutating a package variable in this constructor, which doesn't seem correct

ardaguclu replied:
Fully agree about the concern in here.

I doubt that we get any benefit by defining this unexported default config as global variable. Only two places are using this and both of them are in same package(cmd.go). I'd prefer remove `defaultConfigFlags` altogether, what do you think?

liggitt commented on staging/src/k8s.io/kubectl/pkg/cmd/cmd.go:
```diff
@@ -360,7 +362,7 @@ func NewKubectlCommand(o KubectlOptions) *cobra.Command {
 	kubeConfigFlags := o.ConfigFlags
 	if kubeConfigFlags == nil {
-		kubeConfigFlags = defaultConfigFlags
+		kubeConfigFlags = defaultConfigFlags.WithWarningPrinter(o.IOStreams)
```
same comment about mutating the package variable

ardaguclu commented on staging/src/k8s.io/cli-runtime/pkg/genericclioptions/config_flags.go:
```diff
@@ -428,6 +438,12 @@ func (f *ConfigFlags) WithWrapConfigFn(wrapConfigFn func(*rest.Config) *rest.Con
 }
 
+// WithWarningPrinter initializes WarningPrinter with the given IOStreams
+func (f *ConfigFlags) WithWarningPrinter(ioStreams genericiooptions.IOStreams) *ConfigFlags {
```
Although their names and purposes are similar, they don't interact with each other. As you already know `WarningHandler` is to print API Server warnings https://github.com/kubernetes/kubernetes/blob/52353a2e4ac121ccbfc0fc243e5738888286cc96/staging/src/k8s.io/client-go/rest/warnings.go#L144

On the other hand, `WarningPrinter` is standardized way to print warning for kubectl commands(like https://github.com/kubernetes/kubernetes/blob/00c30941260a27e6929aef84c7fdbc8f1508518c/staging/src/k8s.io/kubectl/pkg/cmd/delete/delete.go#L137) and we are passing it to cli-runtime.

ardaguclu committed:
Remove defaultConfigFlag global variable

ardaguclu commented:
/test pull-kubernetes-e2e-kind

ardaguclu commented:
Failure is not related to the changes in this PR;
/test pull-kubernetes-e2e-gce

liggitt commented on staging/src/k8s.io/kubectl/pkg/cmd/cmd.go:
```diff
@@ -96,15 +96,15 @@ type KubectlOptions struct {
 	genericiooptions.IOStreams
 }
 
-var defaultConfigFlags = genericclioptions.NewConfigFlags(true).WithDeprecatedPasswordFlag().WithDiscoveryBurst(300).WithDiscoveryQPS(50.0)
```
rather than repeating this in two places, make this a private function we can call if needed

```go
func defaultConfigFlags() *genericclioptions.ConfigFlags {
 
 return genericclioptions.NewConfigFlags(true).WithDeprecatedPasswordFlag().WithDiscoveryBurst(300).WithDiscoveryQPS(50.0)
}
```

liggitt commented on staging/src/k8s.io/kubectl/pkg/cmd/cmd.go:
```diff
@@ -96,15 +96,15 @@ type KubectlOptions struct {
 	return NewDefaultKubectlCommandWithArgs(KubectlOptions{
 		PluginHandler: NewDefaultPluginHandler(plugin.ValidPluginFilenamePrefixes),
 		Arguments:     os.Args,
 		ConfigFlags:   defaultConfigFlags,
```
```suggestion
		ConfigFlags:   defaultConfigFlags().WithWarningPrinter(ioStreams),
```

liggitt commented on staging/src/k8s.io/kubectl/pkg/cmd/cmd.go:
```diff
@@ -360,7 +360,7 @@ func NewKubectlCommand(o KubectlOptions) *cobra.Command {
 	kubeConfigFlags := o.ConfigFlags
 	if kubeConfigFlags == nil {
-		kubeConfigFlags = defaultConfigFlags
+		kubeConfigFlags = genericclioptions.NewConfigFlags(true).WithDeprecatedPasswordFlag().WithDiscoveryBurst(300).WithDiscoveryQPS(50.0).WithWarningPrinter(o.IOStreams)
```
```suggestion
		kubeConfigFlags = defaultConfigFlags().WithWarningPrinter(o.IOStreams)
```

ardaguclu committed:
Move default config flags initialization into function

ardaguclu committed:
Skip warning for versions of same group/resource

ardaguclu commented:
@liggitt could you please take a look at this one when you get a chance?. Thanks.

ardaguclu committed:
Run update-vendor

ardaguclu commented:
@liggitt if you get a chance, could you PTAL?. It would be great if we can land this in 1.29. Thanks.

liggitt commented on staging/src/k8s.io/client-go/restmapper/shortcut.go:
```diff
@@ -145,16 +148,33 @@ func (e shortcutExpander) expandResourceShortcut(resource schema.GroupVersionRes
+						continue
+					}
+					if e.warningHandler != nil {
+						e.warningHandler(fmt.Sprintf("ambiguous resource %q for %s is skipped", item.LongForm, resource))
```
I'd suggest something like this to read better:

```go
fmt.Sprintf("short name %q could also match lower priority resource %s.%s", resource.Resource, item.LongForm.Resource, item.LongForm.Group)
```

ardaguclu replied:
Thanks for review.

ardaguclu committed:
Warn only once when there are multiple versions registered for ambiguous resource

ardaguclu commented on staging/src/k8s.io/client-go/restmapper/shortcut.go:
```diff
@@ -145,16 +148,33 @@ func (e shortcutExpander) expandResourceShortcut(resource schema.GroupVersionRes
-				resource.Resource = item.LongForm.Resource
-				resource.Group = item.LongForm.Group
-				return resource
+				if found {
+					if item.LongForm.Group == rsc.Group && item.LongForm.Resource == rsc.Resource {
```
Good idea to prevent noises for a resource with multiple versions registered. Updated.

ardaguclu committed:
Apply gocritic review

ardaguclu commented:
Unrelated failure;
/test pull-kubernetes-node-e2e-containerd

ardaguclu commented:
/test pull-kubernetes-node-e2e-containerd

ardaguclu commented:
@liggitt would you mind reviewing one last round, when you have a chance?. Thanks.

ardaguclu commented:
@liggitt sorry for pinging regularly. I do this to ensure that this PR doesn't slip through the cracks.

liggitt commented on staging/src/k8s.io/client-go/restmapper/shortcut_test.go:
```diff
@@ -255,6 +259,148 @@ func TestKindForWithNewCRDs(t *testing.T) {
 	}
 }
 
+func TestWarnAmbigious(t *testing.T) {
```
can you add a testcase that includes multiple ambiguous groups with multiple versions, that captures all the warnings that would be output in that case, to exercise the deduplication by version and group?

ardaguclu replied:
@liggitt I've added test cases for multiple-resource with multiple-versions. Could you PTAL?.

liggitt commented on staging/src/k8s.io/client-go/restmapper/shortcut_test.go:
```diff
@@ -255,6 +259,148 @@ func TestKindForWithNewCRDs(t *testing.T) {
+		warnedCount := 0
+		mapper := NewShortcutExpander(&fakeRESTMapper{}, ds, func(a string) {
+			warnedCount += 1
+			if warnedCount > 1 {
+				t.Fatalf("warning fired more than expected")
+			}
```
I think we actually want a testcase that exercises multiple warnings since that's a scenario this PR introduces

liggitt replied:
you could append to a list of warnings encountered, then compare the list received to a list of warnings expected

ardaguclu replied:
I've updated test function to handle multiple warnings. Let me know what you think?. Thanks.

ardaguclu replied:
@liggitt would you mind taking a look when you have a chance?. Thanks.

ardaguclu committed:
Add multi-resource multi-version ambiguity unit test

ardaguclu commented:
Flaky test failure;
/test pull-kubernetes-unit

liggitt commented:
cc @seans3  @aojea 

```
{Failed;  === RUN   TestWebSocketClient_ProtocolVersions
E1002 13:45:45.706699   64713 v2.go:150] next reader: websocket: close 1006 (abnormal closure): unexpected EOF
E1002 13:45:45.707990   64713 v2.go:150] next reader: websocket: close 1006 (abnormal closure): unexpected EOF
E1002 13:45:50.706933   64713 websocket.go:478] Websocket Ping failed: write tcp 127.0.0.1:36094->127.0.0.1:42119: use of closed network connection
E1002 13:45:50.709236   64713 websocket.go:478] Websocket Ping failed: write tcp 127.0.0.1:36098->127.0.0.1:42119: use of closed network connection
E1002 13:45:55.709664   64713 websocket.go:478] Websocket Ping failed: write tcp 127.0.0.1:36094->127.0.0.1:42119: use of closed network connection
E1002 13:45:55.709705   64713 websocket.go:478] Websocket Ping failed: write tcp 127.0.0.1:36098->127.0.0.1:42119: use of closed network connection
E1002 13:45:55.709822   64713 websocket.go:478] Websocket Ping failed: write tcp 127.0.0.1:36104->127.0.0.1:42119: write: broken pipe
E1002 13:46:00.706908   64713 websocket.go:478] Websocket Ping failed: write tcp 127.0.0.1:36094->127.0.0.1:42119: use of closed network connection
E1002 13:46:00.709127   64713 websocket.go:478] Websocket Ping failed: write tcp 127.0.0.1:36098->127.0.0.1:42119: use of closed network connection
E1002 13:46:00.710298   64713 websocket.go:478] Websocket Ping failed: write tcp 127.0.0.1:36104->127.0.0.1:42119: write: broken pipe
E1002 13:46:05.706953   64713 websocket.go:478] Websocket Ping failed: write tcp 127.0.0.1:36094->127.0.0.1:42119: use of closed network connection
E1002 13:46:05.709183   64713 websocket.go:478] Websocket Ping failed: write tcp 127.0.0.1:36098->127.0.0.1:42119: use of closed network connection
E1002 13:46:05.709273   64713 websocket.go:478] Websocket Ping failed: write tcp 127.0.0.1:36104->127.0.0.1:42119: write: broken pipe
E1002 13:46:10.707054   64713 websocket.go:478] Websocket Ping failed: write tcp 127.0.0.1:36094->127.0.0.1:42119: use of closed network connection
E1002 13:46:10.708238   64713 websocket.go:478] Websocket Ping failed: write tcp 127.0.0.1:36098->127.0.0.1:42119: use of closed network connection
E1002 13:46:10.709404   64713 websocket.go:478] Websocket Ping failed: write tcp 127.0.0.1:36104->127.0.0.1:42119: write: broken pipe
E1002 13:46:15.707379   64713 websocket.go:478] Websocket Ping failed: write tcp 127.0.0.1:36094->127.0.0.1:42119: use of closed network connection
    websocket_test.go:752: expect stream to be closed after connection is closed.
```
